django/contrib/contenttypes/fields.py

```diff
@@ -194,7 +207,12 @@ def get_prefetch_queryset(self, instances, queryset=None):
         for ct_id, fkeys in fk_dict.items():
             instance = instance_dict[ct_id]
             ct = self.get_content_type(id=ct_id, using=instance._state.db)
-            ret_val.extend(ct.get_all_objects_for_this_type(pk__in=fkeys))
+            if ct_id in custom_queryset_dict:
+                # If the custom queryset is already provided, then directly return that,
+                # otherwise get the data from the database
+                ret_val.extend(list(custom_queryset_dict[ct_id]))
```

Comment:
Why we evaluate the queryset? Couldn't we keep the `QuerySet` object?
```suggestion
                ret_val.extend(custom_queryset_dict[ct_id])
```